datafusion/functions/src/regex/regexpsubstr.rs

```diff
@@ -0,0 +1,531 @@
+                    TypeSignature::Exact(vec![Utf8, Utf8, Int32, Int32]),
+                    TypeSignature::Exact(vec![LargeUtf8, LargeUtf8, Int32, Int32]),
+                    TypeSignature::Exact(vec![Utf8, Utf8, Int32, Int32, Utf8]),
+                    TypeSignature::Exact(vec![
+                        LargeUtf8, LargeUtf8, Int32, Int32, LargeUtf8,
+                    ]),
+                    TypeSignature::Exact(vec![Utf8, Utf8, Int32, Int32, Utf8, Int32]),
+                    TypeSignature::Exact(vec![
+                        LargeUtf8, LargeUtf8, Int32, Int32, LargeUtf8, Int32,
+                    ]),
+                ],
+                Volatility::Immutable,
+            ),
+        }
+    }
+}
+
+impl ScalarUDFImpl for RegexpSubstrFunc {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn name(&self) -> &str {
+        "regexp_substr"
+    }
+
+    fn signature(&self) -> &Signature {
+        &self.signature
+    }
+
+    fn return_type(&self, arg_types: &[DataType]) -> Result<DataType> {
+        Ok(arg_types[0].clone())
+    }
+
+    fn invoke_with_args(&self, args: ScalarFunctionArgs) -> Result<ColumnarValue> {
+        let len = args
+            .args
+            .iter()
+            .fold(Option::<usize>::None, |acc, arg| match arg {
+                ColumnarValue::Scalar(_) => acc,
+                ColumnarValue::Array(a) => Some(a.len()),
+            });
+
+        let is_scalar = len.is_none();
+        let inferred_length = len.unwrap_or(1);
+        let args = args
+            .args
+            .iter()
+            .map(|arg| arg.to_array(inferred_length))
+            .collect::<Result<Vec<_>>>()?;
+
+        let result = regexp_match_func(&args);
+        if is_scalar {
+            // If all inputs are scalar, keeps output as scalar
+            let result = result.and_then(|arr| ScalarValue::try_from_array(&arr, 0));
+            result.map(ColumnarValue::Scalar)
+        } else {
+            result.map(ColumnarValue::Array)
+        }
+    }
+
+    fn documentation(&self) -> Option<&Documentation> {
+        Some(get_regexp_substr_doc())
+    }
+}
+
+static DOCUMENTATION: OnceLock<Documentation> = OnceLock::new();
+
+fn get_regexp_substr_doc() -> &'static Documentation {
+    DOCUMENTATION.get_or_init(|| {
+        Documentation::builder(
+            DOC_SECTION_REGEX,
+            "Returns the substring that matches a [regular expression](https://docs.rs/regex/latest/regex/#syntax) within a string.",
+            "regexp_substr(str, regexp[, position[, occurrence[, flags[, group_num]]]])")
+            .with_sql_example(r#"```sql
+            > select regexp_substr('Köln', '[a-zA-Z]ö[a-zA-Z]{2}');
+            +---------------------------------------------------------+
+            | regexp_substr(Utf8("Köln"),Utf8("[a-zA-Z]ö[a-zA-Z]{2}")) |
+            +---------------------------------------------------------+
+            | [Köln]                                                  |
```

Comment:
here and in other examples we should return string of ?original string datatype (OST). Not list(OST). I have no idea why DF decided to return list ( `DataType::List(Arc::new(Field::new_list_field(other.clone(), true)))` for `regexp_match` but we want to return string. It's correctly exposed in `return_type` function above. So, let's just change the examples.